common/persistence/cassandra/factory.go

```diff
@@ -43,9 +44,14 @@ type (
 		clusterName string
 		logger      log.Logger
 		session     gocql.Session
+		shardLock   locks.IDMutex
 	}
 )
 
+const (
+	shardLockShards = 4
```

Comment:
Haha.. two levels of shards: the history shards get hashed into "shards" inside IDMutex, purely to reduce contention on the internal mutexes in IDMutex. I'm not at all sure it's even necessary, but the code was already there, so I left it when I was changing the interface.

A better name might help, maybe `shardLockIDMutexShards`? I can add a comment also.

Reply:
Why not do 1 to 1 map (each history shard got its own idMutex)?

Reply:
What you're suggesting is one "idshards", not n "idshards" (using "idshards" to refer to shards of IDMutex). Each shard has a map in it, and it's the mutex protecting that map that we want to reduce contention on. That's fine, we can do that. The "idshard" functionality was already there, though, and I think it would help in some situations (though we should benchmark it to make sure there aren't issues with false sharing, etc).

So I'll say three options:

1. Leave the number of "idshards" as some small constant.
2. Set the number of "idshards" to one for this use, but leave the sharding functionality in IDMutex.
3. Remove all sharding functionality from IDMutex, force one shard for all use cases.

I went with 1 so far. I do appreciate simplifying code, but I'm worried someone will just end up adding the sharding back in 1-2 years. :shrug: 

(Also, if we're going for simplification, we could remove the non-"cleanup" code path as well)

Reply:
I would vote for option 3. Using a RWMutex should be enough as there rate we need to add to the map is low.